### PR TITLE
Refactor sr-api to not depend on client anymore

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

**Comment by tomaka** [comment]
I know we want to go fast, but I'm not a fan of marking a PR that touches 2500 lines of code as "insubstantial" and without any explanation.

**Comment by bkchr** [comment]
That was by mistake and I already have removed that (before your comment :P).

File: core/sr-api/proc-macro/src/impl_runtime_apis.rs [hunk 1]
```diff
@@ -657,5 +655,8 @@ pub fn impl_runtime_apis_impl(input: proc_macro::TokenStream) -> proc_macro::Tok
 
 			#wasm_interface
 		}
-	).into()
+	);
+	// println!("{}", res);
```
Comment (tomusdrw) [inline core/sr-api/proc-macro/src/impl_runtime_apis.rs]:
```suggestion
```

File: core/sr-api/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
```
Comment (tomusdrw) [inline core/sr-api/src/lib.rs]: It's a new file, right? Should be at `2019`?

File: core/block-builder/src/lib.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
```
[the remark this answers is not in the document]
Comment (bkchr) [reply core/block-builder/src/lib.rs]: Probably some copy paste error by me.

Comment (bkchr) [inline core/block-builder/src/lib.rs]:
```suggestion
// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```

File: core/client/src/client.rs [hunk 4]
```diff
@@ -53,11 +53,11 @@ use consensus::{
 };
 use header_metadata::{HeaderMetadata, CachedHeaderMetadata};
 
+use sr_api::{CallRuntimeAt, ConstructRuntimeApi, Core as CoreApi, ProofRecorder, InitializeBlock};
+
```
Comment (bkchr) [inline core/client/src/client.rs]:
```suggestion
```

File: core/consensus/common/src/error.rs [hunk 5]
```diff
@@ -35,8 +35,8 @@ pub enum Error {
 	#[display(fmt="Timer error: {}", _0)]
 	FaultyTimer(std::io::Error),
 	/// Error while working with inherent data.
-	#[display(fmt="InherentData error: {}", _0)]
-	InherentData(String),
+	#[display(fmt="InherentData error: {:?}", _0)]
```
Comment (bkchr) [inline core/consensus/common/src/error.rs]:
```suggestion
	#[display(fmt="InherentData error: {}", _0)]
```

[the remark this answers is not in the document]
Comment (bkchr) [reply core/consensus/common/src/error.rs]: It implements it now, but not at the point where I made this change.

File: core/consensus/pow/src/lib.rs [hunk 6]
```diff
@@ -74,8 +74,8 @@ pub enum Error<B: BlockT> {
 	BestHashSelectChain(ConsensusError),
 	#[display(fmt = "Error with block built on {:?}: {:?}", _0, _1)]
 	BlockBuiltError(B::Hash, ConsensusError),
-	#[display(fmt = "Creating inherents failed: {}", _0)]
-	CreateInherents(RuntimeString),
+	#[display(fmt = "Creating inherents failed: {:?}", _0)]
```
Comment (bkchr) [inline core/consensus/pow/src/lib.rs]:
```suggestion
	#[display(fmt = "Creating inherents failed: {}", _0)]
```

File: core/sr-api/src/lib.rs (hunk 2 above)
Comment (bkchr) [inline core/sr-api/src/lib.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: core/sr-api/proc-macro/src/impl_runtime_apis.rs [hunk 7]
```diff
@@ -639,7 +637,7 @@ pub fn impl_runtime_apis_impl(input: proc_macro::TokenStream) -> proc_macro::Tok
 	let wasm_interface = unwrap_or_error(generate_wasm_interface(&api_impls));
 	let api_impls_for_runtime_api = unwrap_or_error(generate_api_impl_for_runtime_api(&api_impls));
 
-	quote!(
+	let res = quote!(
```
Comment (bkchr) [inline core/sr-api/proc-macro/src/impl_runtime_apis.rs]:
```suggestion
	quote!(
```

File: core/sr-api/proc-macro/src/impl_runtime_apis.rs [hunk 8]
```diff
@@ -657,5 +655,8 @@ pub fn impl_runtime_apis_impl(input: proc_macro::TokenStream) -> proc_macro::Tok
 
 			#wasm_interface
 		}
-	).into()
+	);
+	// println!("{}", res);
+
```
Comment (bkchr) [inline core/sr-api/proc-macro/src/impl_runtime_apis.rs]:
```suggestion
```

File: core/sr-api/proc-macro/src/impl_runtime_apis.rs [hunk 9]
```diff
@@ -657,5 +655,8 @@ pub fn impl_runtime_apis_impl(input: proc_macro::TokenStream) -> proc_macro::Tok
...
+	res.into()
```
Comment (bkchr) [inline core/sr-api/proc-macro/src/impl_runtime_apis.rs]:
```suggestion
```

File: core/sr-api/proc-macro/src/impl_runtime_apis.rs [hunk 10]
```diff
@@ -657,5 +655,6 @@ pub fn impl_runtime_apis_impl(input: proc_macro::TokenStream) -> proc_macro::Tok
 
 			#wasm_interface
 		}
-	).into()
+	);
```
Comment (bkchr) [inline core/sr-api/proc-macro/src/impl_runtime_apis.rs]:
```suggestion
	).into()
```

**Pull request merged**